impeller/entity/geometry/stroke_path_geometry.cc

```diff
@@ -242,14 +242,99 @@ StrokePathGeometry::CreateSolidStrokeVertices(
 
-  auto compute_offset = [&polyline, &offset, &previous_offset,
-                         &stroke_width](size_t point_i) {
+  // Computes offset by calculating the direction from point_i - 1 to point_i if
+  // point_i is within `contour_start_point_i` and `contour_end_point_i`;
+  // Otherwise, it uses direction from contour.
+  auto compute_offset = [&polyline, &offset, &previous_offset, &stroke_width](
+                            const size_t point_i,
+                            const size_t contour_start_point_i,
+                            const size_t contour_end_point_i,
+                            const Path::PolylineContour& contour) {
+    Point direction;
+    if (point_i >= contour_end_point_i) {
+      direction = contour.end_direction;
+    } else if (point_i <= contour_start_point_i) {
+      direction = -contour.start_direction;
+    } else {
+      direction =
+          (polyline.points[point_i] - polyline.points[point_i - 1]).Normalize();
+    }
     previous_offset = offset;
-    Point direction =
-        (polyline.points[point_i] - polyline.points[point_i - 1]).Normalize();
     offset = Vector2{-direction.y, direction.x} * stroke_width * 0.5;
   };
 
+  auto add_vertices_for_linear_compoent =
+      [&vtx_builder, &offset, &previous_offset, &vtx, &polyline,
+       &compute_offset, scaled_miter_limit, scale, &join_proc](
+          const size_t component_start_index, const size_t component_end_index,
+          const size_t contour_start_point_i, const size_t contour_end_point_i,
+          const Path::PolylineContour& contour) {
+        auto is_last_component =
+            component_start_index ==
+            contour.components.back().component_start_index;
+
+        for (size_t point_i = component_start_index;
+             point_i < component_end_index; point_i++) {
+          auto is_end_of_component = point_i == component_end_index - 1;
+          vtx.position = polyline.points[point_i] + offset;
+          vtx_builder.AppendVertex(vtx);
+          vtx.position = polyline.points[point_i] - offset;
+          vtx_builder.AppendVertex(vtx);
+
+          // For line components, two additional points need to be appended
+          // prior to appending a join connecting the next component.
+          vtx.position = polyline.points[point_i + 1] + offset;
+          vtx_builder.AppendVertex(vtx);
+          vtx.position = polyline.points[point_i + 1] - offset;
+          vtx_builder.AppendVertex(vtx);
+
+          compute_offset(point_i + 2, contour_start_point_i,
+                         contour_end_point_i, contour);
+          if (!is_last_component && is_end_of_component) {
+            // Generate join from the current line to the next line.
+            join_proc(vtx_builder, polyline.points[point_i + 1],
+                      previous_offset, offset, scaled_miter_limit, scale);
+          }
+        }
+      };
+
+  auto add_vertices_for_curve_compoent =
```

Comment:
There are still one more issue with curve component. the overdraw from the few start points will draw outside of the start point. I would like to address this in a later pr.
![image](https://github.com/flutter/engine/assets/47866232/448788c1-0aae-445f-8ed4-719991e20dc4)

The start of the curve is not exactly horizontal due to this issue.